### ISSUE TITLE
[with-firebase-authentication] Can't use Firebase Admin SDK for anything other than verifying the token

### ISSUE THREAD

**Issue by tomthornton** [vercel/next.js#14139]
# Bug report

## Describe the bug

There's a problem with the credential in initializing the firebase admin app. It works perfectly with admin.verifyToken, but any other method I try to use yields an error of

`Error: Credential implementation provided to initializeApp() via the "credential" property failed to fetch a valid Google OAuth2 access token with the following error: "Error fetching access token: invalid_grant (Invalid grant: account not found)". There are two likely causes: (1) your server time is not properly synced or (2) your certificate key file has been revoked. To solve (1), re-sync the time on your server. To solve (2), make sure the key ID for your key file is still present at https://console.firebase.google.com/iam-admin/serviceaccounts/project. If not, generate a new key file at https://console.firebase.google.com/project/_/settings/serviceaccounts/adminsdk.`

I've tried everything. Is there no way to use the firebase admin SDK with anything else in NextJS? 

## To Reproduce

firebaseAdmin.js

import * as admin from 'firebase-admin'

export const createUser = (email) => {
  const firebasePrivateKey = process.env.FIREBASE_PRIVATE_KEY

  if (!admin.apps.length) {
    admin.initializeApp({
      credential: admin.credential.cert({
        projectId: process.env.NEXT_PUBLIC_FIREBASE_PROJECT_ID,
        clientEmail: process.env.FIREBASE_CLIENT_EMAIL,
        // https://stackoverflow.com/a/41044630/1332513
        privateKey: firebasePrivateKey.replace(/\\n/g, '\n'),
      }),
      databaseURL: process.env.NEXT_PUBLIC_FIREBASE_DATABASE_URL,
    })
  } 

  return admin.auth().createUser({
    email: 'example@test.com',
    password: 'password'
  })
}

export const verifyIdToken = (token) => {
  const firebasePrivateKey = process.env.FIREBASE_PRIVATE_KEY
  if (!admin.apps.length) {
    admin.initializeApp({
      credential: admin.credential.cert({
        projectId: process.env.NEXT_PUBLIC_FIREBASE_PROJECT_ID,
        clientEmail: process.env.FIREBASE_CLIENT_EMAIL,
        // https://stackoverflow.com/a/41044630/1332513
        privateKey: firebasePrivateKey.replace(/\\n/g, '\n'),
      }),
      databaseURL: process.env.NEXT_PUBLIC_FIREBASE_DATABASE_URL,
    })
  }

  return admin
    .auth()
    .verifyIdToken(token)
    .catch((error) => {
      throw error
    })
}

Try to run createUser from an API endpoint

## Expected behavior

It should just work. I need a custom flow for creating my users.

## System information
- Mac.OS
- Node '12.16.1'
- NextJS - latest

**Comment by RiChrisMurphy** [comment]
did you ever solve this issue? I'm having the same one.

**Comment by RiChrisMurphy** [comment]
Well, I ended up solving my issue by using the entire service-account.json instead of only the private key in the admin.credential.cert( ... ). It seems to work on localhost with just the private key (without the rest of the service account properties), which threw me off when I moved to other environments.

**Comment by tomthornton** [comment]
> Well, I ended up solving my issue by using the entire service-account.json instead of only the private key in the admin.credential.cert( ... ). It seems to work on localhost with just the private key (without the rest of the service account properties), which threw me off when I moved to other environments.

Yep I got it to work the same way. Sorry, I didn't see this earlier. Is there a way to add the answer and close this issue?

**Comment by Ey-Jay** [comment]
Do you mean you passed the JSON file path, stored somewhere in your server? Or did you just add the missing properties to admin.credential.cert({...props})?

I just tried passing the missing properties but I still get the same error.

**Comment by Ey-Jay** [comment]
> Well, I ended up solving my issue by using the entire service-account.json instead of only the private key in the admin.credential.cert( ... ). It seems to work on localhost with just the private key (without the rest of the service account properties), which threw me off when I moved to other environments.

Do you mean you passed the JSON file path, stored somewhere in your server? Or did you just add the missing properties to admin.credential.cert({...props})?

I just tried passing the missing properties but I still get the same error.

**Comment by GabrielNBDS** [comment]
How I solved: I deployed it to digital ocean vps and deployed it as a node app.

**Comment by modderzabbasi** [comment]
Is there an example on how to solve this issue in vercel?

**Comment by kedarguy** [comment]
I got it working by saving the private key directly in the file.
```
const privateKey = '-----BEGIN PRIVATE KEY-----\nXXXXXXX-----END PRIVATE KEY-----\n

function initializeFirebaseAdmin() {
  if (!admin.apps.length) {
    return admin.initializeApp({
      credential: admin.credential.cert({
        projectId: process.env.NEXT_PUBLIC_FIREBASE_PROJECT_ID,
        clientEmail: process.env.FIREBASE_CLIENT_EMAIL,
        privateKey: privateKey.replace(/\\n/g, '\n'),
      }),
      databaseURL: process.env.NEXT_PUBLIC_FIREBASE_DATABASE,
    });
  }
  return admin.apps[0];
}
```


So I think the issue is adding the firebase private key as an ENV var on Vercel.

It is not ideal to hardcode the private key, so I would be happy to hear if someone got it to work using ENV vars on vercel.

**Comment by sinhaGuild** [comment]
Try adding a [prefix](https://nextjs.org/docs/basic-features/environment-variables#exposing-environment-variables-to-the-browser) of `NEXT_PUBLIC_` to all your environment variables.

`.env.local` will look like this

```sh
NEXT_PUBLIC_FIREBASE_API_KEY=''
NEXT_PUBLIC_FIREBASE_AUTHD_OMAIN=''
NEXT_PUBLIC_FIREBASE_PROJECT_ID=''
NEXT_PUBLIC_FIREBASE_STORAGE_BUCKET=''
NEXT_PUBLIC_FIREBASE_MESSAGING_SENDER_ID=''
NEXT_PUBLIC_FIREBASE_APP_ID=''
NEXT_PUBLIC_FIREBASE_MEASUREMENT_ID=''
```

and then referencing it in your `*.ts/tsx/js/jsx` like so.

```js
const firebaseConfig = {
    apiKey: process.env.NEXT_PUBLIC_FIREBASE_API_KEY,
    authDomain: process.env.NEXT_PUBLIC_FIREBASE_AUTH_DOMAIN,
    projectId: process.env.NEXT_PUBLIC_FIREBASE_PROJECT_ID,
    storageBucket: process.env.NEXT_PUBLIC_FIREBASE_STORAGE_BUCKET,
    messagingSenderId: process.env.NEXT_PUBLIC_FIREBASE_MESSAGING_SENDER_ID,
    appId: process.env.NEXT_PUBLIC_FIREBASE_APP_ID,
    measurementId: process.env.NEXT_PUBLIC_FIREBASE_MEASUREMENT_ID,
  };
```

**Comment by laurentvergnaud** [comment]
> NEXT_PUBLIC_

For a Firebase Admin Private Key ?!
